Mathlib/Order/Directed.lean

```diff
@@ -321,3 +325,57 @@ instance (priority := 100) OrderTop.to_isDirected_le [LE α] [OrderTop α] : IsD
 -- see Note [lower instance priority]
 instance (priority := 100) OrderBot.to_isDirected_ge [LE α] [OrderBot α] : IsDirected α (· ≥ ·) :=
   ⟨fun _ _ => ⟨⊥, bot_le _, bot_le _⟩⟩
+
+namespace DirectedOn
+
+section Pi
+
+variable {ι : Type*} {α : ι → Type*}
+
+variable (r : (i : ι) →  α i →  α i → Prop)
+
+lemma proj {d : Set (Π i, α i)} (hd : DirectedOn (fun x y => ∀ i, r i (x i)  (y i)) d) (i : ι) :
+    DirectedOn (r i) ((fun a => a i) '' d) :=
+  DirectedOn.mono_comp (fun _ _ h => h) (mono hd fun ⦃_ _⦄ h ↦ h i)
+
+lemma pi {d : (i : ι) → Set (α i)} (hd : ∀ (i : ι), DirectedOn (r i) (d i)) :
+    DirectedOn (fun x y => ∀ i, r i (x i) (y i)) (Set.pi  Set.univ d) := by
+  intro a ha b hb
+  choose f hf using fun i => hd i (a i) (ha i trivial) (b i) (hb i trivial)
+  simpa [Pi.le_def, ← forall_and] using ⟨f, fun i =>
+    ⟨by simp_all only [Set.mem_pi, Set.mem_univ, forall_const],
+       by constructor <;> (simp_all only [Set.mem_pi, Set.mem_univ, forall_const])⟩⟩
```

Comment:
Here's a nice trick: the choose tactic allows you to break out `hf` into different sub-hypotheses (well actually it doesn't distinguish between `f` and `hf` either, it justs splits as long as you ask for), so you can avoid having a bunch of "and"s under a forall. That makes the rest of the proof look much cleaner:
```suggestion
  choose f hfd haf hbf using fun i => hd i (a i) (ha i trivial) (b i) (hb i trivial)
  exact ⟨f, fun i _ => hfd i, haf, hbf⟩
```